core/cache/modelsummarywatcher_test.go

```diff
@@ -0,0 +1,578 @@
+			ApplicationCount: 1,
+		},
+	})
+}
+
+func (s *modelSummaryWatcherSuite) TestChangesToOneModelCoalesced(c *gc.C) {
+	watcher := s.controller.WatchAllModels()
+	defer workertest.CleanKill(c, watcher)
+
+	changes := watcher.Changes()
+	// discard the intial event
+	_ = s.next(c, changes)
+
+	s.ProcessChange(c, cache.RemoveUnit{
+		ModelUUID: "model-1-uuid",
+		Name:      "magic/0",
+	}, s.events)
+	s.ProcessChange(c, cache.RemoveApplication{
+		ModelUUID: "model-1-uuid",
+		Name:      "magic",
+	}, s.events)
+	s.ProcessChange(c, cache.RemoveMachine{
+		ModelUUID: "model-1-uuid",
+		Id:        "0",
+	}, s.events)
+	s.ProcessChange(c, cache.ApplicationChange{
+		ModelUUID: "model-2-uuid",
+		Name:      "foo",
+		Life:      life.Alive,
+	}, s.events)
+
+	// Make sure all the published summary events have been handled.
+	// We know that the events are consumed in the order that they are
+	// published, and that any publishing will take place during the
+	// above ProcessChange calls above, so we only need to wait on the
+	// last publish event from model 2.
+	cache.WaitForModelSummaryHandled(c, s.controller, "model-2-uuid")
+
+	update := s.next(c, changes, "model-2-uuid")
+	c.Assert(update, jc.DeepEquals, []cache.ModelSummary{
+		{
+			UUID:      "model-1-uuid",
+			Namespace: "test-admin",
+			Name:      "model-1",
+			Admins:    []string{"test-admin"},
+			Status:    cache.StatusGreen,
+		}, {
+			UUID:             "model-2-uuid",
+			Namespace:        "bob",
+			Name:             "model-2",
+			Admins:           []string{"bob"},
+			Status:           cache.StatusGreen,
+			ApplicationCount: 1,
+		},
+	})
+}
+
+func (s *modelSummaryWatcherSuite) TestUpdatesThatDontChangeSummary(c *gc.C) {
+	watcher := s.controller.WatchAllModels()
+	defer workertest.CleanKill(c, watcher)
+
+	changes := watcher.Changes()
+	// discard the intial event
+	_ = s.next(c, changes)
+
+	modelUpdate := cache.ModelChange{
+		ModelUUID: "new-model-uuid",
+		Name:      "new-model",
+		Life:      life.Alive,
+		Owner:     "mary",
+		UserPermissions: map[string]permission.Access{
+			"bob":  permission.ReadAccess,
+			"mary": permission.AdminAccess,
+		},
+	}
+	s.ProcessChange(c, modelUpdate, s.events)
+
+	update := s.next(c, changes)
+	c.Assert(update, jc.DeepEquals, []cache.ModelSummary{
+		{
+			UUID:      "new-model-uuid",
+			Namespace: "mary",
+			Name:      "new-model",
+			Admins:    []string{"mary"},
+			Status:    cache.StatusGreen,
+		},
+	})
+
+	// Now we send the same model update, the hash shouldn't change
+	// so it shouldn't generate a new event.
+	s.ProcessChange(c, modelUpdate, s.events)
+	// We send another event after it so we aren't waiting for something
+	// to not happen, which makes tests slower. Here we add an application
+	// to an existing model to force an update.
+	s.ProcessChange(c, cache.ApplicationChange{
+		ModelUUID: "model-2-uuid",
+		Name:      "foo",
+		Life:      life.Alive,
+	}, s.events)
+
+	update = s.next(c, changes, "model-2-uuid")
```

Comment:
did you run this test with the hash dedup skipped, and saw that the test actually broke when the event triggered?
Calling ProcessChange must overwrite the .lastSummaryPublished, which means this next is waiting for that publish to finish. So Coalescing events would also give this result, wouldn't it? I guess your thought is that coalescing would mean that we would see both a model-2 change and a new-model-uuid change.

Reply:
Yes. Internally it is coalescing the changes, so if there was another change for the first model, we'd see it.